### PR TITLE
Add logical plan support for aggregate expressions with filters (and upgrade to sqlparser 0.23)

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/2214

 # Rationale for this change

Adds support to the SQL query planner and logical plan for aggregate expressions with filters e.g. `aggr_expr(foo) FILTER (WHERE ...)`

# What changes are included in this PR?

- `Expr::AggregateFunction` and `Expr::AggregateUDF` have a new `filter` attribute
- SQL parser updated to populate this
- Physical plan and proto throw an error if this filter is populated

# Are there any user-facing changes?

Yes, `Expr` API change

### CODE REVIEW

**Comment by andygrove** [comment]
cc @ayushdg

File: datafusion/expr/src/expr.rs [hunk 1]
```diff
@@ -231,6 +231,8 @@ pub enum Expr {
         args: Vec<Expr>,
         /// Whether this is a DISTINCT aggregation or not
         distinct: bool,
+        /// Optional filter
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:234]:
```suggestion
        /// Optional filter applied prior to aggregating
```

File: datafusion/expr/src/expr.rs [hunk 2]
```diff
@@ -251,6 +253,8 @@ pub enum Expr {
         fun: Arc<AggregateUDF>,
         /// List of expressions to feed to the functions as arguments
         args: Vec<Expr>,
+        /// Optional filter
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:256]:
```suggestion
        /// Optional filter applied prior to aggregating
```

File: datafusion/proto/src/lib.rs [hunk 3]
```diff
@@ -1009,6 +1012,7 @@ mod roundtrip_tests {
         let test_expr = Expr::AggregateUDF {
             fun: Arc::new(dummy_agg.clone()),
             args: vec![lit(1.0_f64)],
+            filter: Some(Box::new(lit(true))),
```
Comment (alamb) [inline datafusion/proto/src/lib.rs:1015]: 👍

File: datafusion/proto/src/to_proto.rs [hunk 4]
```diff
@@ -639,17 +644,23 @@ impl TryFrom<&Expr> for protobuf::LogicalExprNode {
...
+                            fun_name: fun.name.clone(),
+                            args: args.iter().map(|expr| expr.try_into()).collect::<Result<
+                                Vec<_>,
+                                Error,
+                            >>(
+                            )?,
+                            filter: match filter {
+                                Some(e) => Some(Box::new(e.as_ref().try_into()?)),
+                                None => None,
+                            }
```
Comment (alamb) [inline datafusion/proto/src/to_proto.rs:660]: I think you can write this kind of thing somewhat more concisely via

```suggestion
                            filter: filter
                              .map(|filter| Box::new(e.as_ref().try_into()))
                              .transpose()?
```

File: datafusion/sql/src/planner.rs [hunk 5]
```diff
@@ -2053,6 +2053,16 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 Ok(Expr::ScalarFunction { fun, args })
             }
 
+            SQLExpr::AggregateExpressionWithFilter { expr, filter } => {
+                match self.sql_expr_to_logical_expr(*expr, schema, ctes)? {
+                    Expr::AggregateFunction {
+                        fun, args, distinct, ..
+                    } =>  Ok(Expr::AggregateFunction { fun, args, distinct, filter: Some(Box::new(self.sql_expr_to_logical_expr(*filter, schema, ctes)?)) }),
+                    _ => Err(DataFusionError::Internal("".to_string()))
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:2061]: The error message here appears to be missing

**Review by alamb — COMMENTED** [review]
Thanks @andygrove 

Looks good to me -- I had some comments, but nothing that needs to be done in this PR (could be done in a follow on).

https://www.postgresql.org/docs/current/sql-expressions.html

> 4.2.7. Aggregate Expressions


Today I learned that postgres aggregates support all sorts of semi-wild things (like an ORDER BY clause)

**Review by alamb — APPROVED** [review]

**Comment by andygrove** [comment]
Thanks for the review @alamb. I fixed the error message.

File: datafusion/proto/src/to_proto.rs (hunk 4 above)
Comment (andygrove) [reply datafusion/proto/src/to_proto.rs:684]: I experimented with this locally but could not get this to work.

**Pull request merged**